### PR TITLE
Adding generic firebase infrastructure

### PR DESCRIPTION
**Pull request by TalShperling**

### CODE REVIEW

File: src/app/services/firebase/firebase.service.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+   */
+  getAllData$<T extends { id: string }>(collection: CollectionType): Observable<T[]> {
+    return this.firestore.collection<T>(collection).valueChanges();
+  }
+
+  getDataById$<T>(collection: CollectionType, id: string): Observable<T> {
+    return this.firestore.doc<T>(`${collection}/${id}`).valueChanges();
+  }
+
+  upsertData<T extends { id: string }>(collection: CollectionType, data: T): Observable<T> {
```
Comment (yarivg) [inline src/app/services/firebase/firebase.service.ts]: rename to upsertDocument

File: src/app/services/firebase/firebase.service.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    return new Observable<T>(observer => {
+      if (!data.id) {
+        data.id = this.firestore.createId();
+      }
+      this.firestore.doc<T>(`${collection}/${data.id}`).set(data).then(() => observer.next(data));
+    });
+  }
+
+  deleteData<T extends { id: string }>(collection: CollectionType, data: T): Observable<void> {
```
Comment (yarivg) [inline src/app/services/firebase/firebase.service.ts]: Rename to deleteDocument

File: src/app/services/users/users.service.ts [hunk 3]
```diff
@@ -2,28 +2,27 @@ import {Injectable} from '@angular/core';
...
+    return this.firebaseService.getAllData$<User>(this.collectionName);
   }
-
-  getUserByID(id: string): Observable<User> {
-    return new Observable<User>(observer => {
-      this.firestore.collection<User>('users').doc(id).get().subscribe(user => {
-        observer.next(user.data() as User);
-      });
-    });
+  getById$(id: string): Observable<User> {
```
Comment (yarivg) [inline src/app/services/users/users.service.ts:18]: Can firebaseservice be an abstract class thatimplements this functions for  a type T for it's users? Don't know if it works in type script.. but think about it. This way you only extend from firebaseService for the simple case.

**Review by yarivg — CHANGES_REQUESTED** [review]
Looks good

File: src/app/services/firebase/firebase.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { TestBed } from '@angular/core/testing';
```
Comment (Ozzorany) [inline src/app/services/firebase/firebase.service.spec.ts]: remove this file. We're not using tests

File: src/app/services/items/items.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { TestBed } from '@angular/core/testing';
```
Comment (Ozzorany) [inline src/app/services/items/items.service.spec.ts]: remove test

File: src/app/services/users/users.service.ts (hunk 3 above)
Comment (TalShperling) [reply src/app/services/users/users.service.ts:18]: I would prefer that the services wouldn't know that they are speaking with a firebase service.
In order to do so I can't define the firebase service as an abstract service.

**Comment by Ozzorany** [comment]
@TalShperling 
Fix conflicts & remove the specs files- we don't have tests and then I'll approve it

**Review by Ozzorany — APPROVED** [review]

**Pull request merged**